server/src/train/trainvehiclelist.cpp

```diff
@@ -89,27 +94,137 @@ TrainVehicleList::TrainVehicleList(Train& train_, std::string_view parentPropert
+      train().updateSpeedMax();
+
+  if(!m_models.empty() && TrainVehicleListTableModel::isListedProperty(property.name()))
+  {
+    ObjectPtr obj = property.object().shared_from_this();
+    const uint32_t rows = static_cast<uint32_t>(items.size());
+    for(uint32_t row = 0; row < rows; row++)
+    {
+      if(items[row] == obj)
+      {
+        for(auto& model : m_models)
+        {
+          model->propertyChanged(property, row);
+          break;
+        }
+      }
+    }
+  }
+}
+
 TableModelPtr TrainVehicleList::getModel()
 {
-  return std::make_shared<RailVehicleListTableModel>(*this);
+  return std::make_shared<TrainVehicleListTableModel>(*this);
 }
 
-bool TrainVehicleList::isListedProperty(std::string_view name)
+void TrainVehicleList::rowCountChanged()
 {
-  return RailVehicleListTableModel::isListedProperty(name);
+  const auto size = items.size();
+  //length.setValueInternal(static_cast<uint32_t>(size));
+  for(auto& model : m_models)
+      model->setRowCount(static_cast<uint32_t>(size));
 }
 
-void TrainVehicleList::propertyChanged(BaseProperty& property)
+void TrainVehicleList::rowsChanged(uint32_t first, uint32_t last)
 {
-  if(property.name() == "lob")
-    train().updateLength();
-  else if(property.name() == "total_weight")
-    train().updateWeight();
-  else if(property.name() == "speed_max")
-    train().updateSpeedMax();
+  for(auto& model : m_models)
+  {
+      model->rowsChanged(first, last);
+  }
+}
+
+void TrainVehicleList::addObject(std::shared_ptr<RailVehicle> vehicle)
+{
+  std::shared_ptr<TrainVehicleListItem> object = getItemFromVehicle(vehicle);
+  if(object)
+    return; //Vehicle is already in this train
+
+  object = std::make_shared<TrainVehicleListItem>(*this, getItemId());
+  object->vehicle.setValueInternal(vehicle);
+
+  items.appendInternal(object);
+  m_propertyChanged.emplace(object.get(), object->propertyChanged.connect(std::bind(&TrainVehicleList::propertyChanged, this, std::placeholders::_1)));
+
+  rowCountChanged();
+  train().vehiclesChanged();
+}
+
+void TrainVehicleList::removeObject(const std::shared_ptr<TrainVehicleListItem> &item)
+{
+  auto it = std::find(items.begin(), items.end(), item);
+  if(it == items.end())
+    return;
+
+  m_propertyChanged[item.get()].disconnect();
+  m_propertyChanged.erase(item.get());
+
+  item->destroy();
+  items.removeInternal(item);
+
+  rowCountChanged();
+
+  uint32_t row = std::distance(items.begin(), it);
+  for(auto& model : m_models)
+  {
+    model->rowRemovedHack(row);
+  }
+
+  train().vehiclesChanged();
 }
 
 Train& TrainVehicleList::train()
 {
   return static_cast<Train&>(parent());
 }
+
+unsigned int TrainVehicleList::getItemId() const
+{
+  uint32_t itemId = 1;
+  while(std::find_if(begin(), end(), [itemId](const auto& it){ return it->itemId() == itemId; }) != end())
+      itemId++;
+  return itemId;
+}
+
+std::shared_ptr<TrainVehicleListItem> TrainVehicleList::getItemFromVehicle(const std::shared_ptr<RailVehicle> vehicle)
```

Comment:
const ref shared_ptr  for argument and return type

for a null return add: `static const std::shared_ptr<TrainVehicleListItem> null;` in the function, the `null` can be returned const ref.